File: backend/app/schemas/ai_helper.py
"""Schemas for the AI Helper bulk-table mini-tool (v1.1: multi-output)."""
from datetime import datetime

from pydantic import BaseModel


class AiHelperOutput(BaseModel):
    """One output column a run writes/edits."""

    column_id: int
    # 'write' (produce new content for the column) | 'edit' (rewrite in place;
    # the column must also be a mapped input so the model sees its content).
    mode: str = "write"
    # JSON key the structured engine routes to this column (auto-slugged from the
    # column name in the UI, editable). Unused by the per_output engine.
    key: str = ""
    # This output's own prompt (per_output engine only; may contain {{vars}}).
    prompt: str = ""


class AiHelperRunCreate(BaseModel):
    # 'structured' (1 AI call/row → JSON keyed by output) | 'per_output'
    # (1 AI call per output, each with its own prompt).
    engine: str = "structured"
    # Structured: the shared base prompt. per_output: optional (each output
    # carries its own prompt); kept for the run summary.
    prompt: str = ""
    prompt_id: int | None = None  # provenance if loaded from the library
    name: str | None = None
    # {var_name: source_column_id} — the columns the prompt(s) read.
    variable_map: dict[str, int] = {}
    # The columns this run writes/edits (≥1; column ids unique).
    outputs: list[AiHelperOutput] = []
    provider_code: str | None = None
    model: str | None = None
    max_output_tokens: int | None = None
    # 'full' | 'first_pct' (word-slice the sliced column's first input_pct%).
    input_scope: str = "full"
    input_pct: int | None = None
    slice_column_id: int | None = None
    # Explicit rows to process (selection / range / filter, resolved by the UI);
    # empty = every row in the table.
    row_ids: list[int] = []


class AiHelperCellRead(BaseModel):
    id: int
    row_id: int
    row_position: int
    column_id: int
    state: str  # pending | done | failed | skipped
    old_value: str | None = None
    new_value: str | None = None
    error: str | None = None


class AiHelperRunRead(BaseModel):
    id: int
    table_id: int
    status: str  # queued | running | cancelled | done | failed
    mode: str  # legacy single-output mode (v1.1 runs: informational)
    engine: str  # structured | per_output
    name: str | None = None
    target_column_id: int | None = None
    total: int
    done: int
    failed: int
    skipped: int
    reverted_at: datetime | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None


class AiHelperRunDetail(AiHelperRunRead):
    prompt: str = ""
    variable_map: dict = {}
    # Effective outputs (synthesized from target_column_id/mode for legacy runs).
    outputs: list[AiHelperOutput] = []
    provider_code: str | None = None
    model: str | None = None
    input_scope: str = "full"
    input_pct: int | None = None
    error: str | None = None
    items: list[AiHelperCellRead] = []
    items_total: int = 0
    items_page: int = 1
    items_page_size: int = 50


class AiHelperPreview(BaseModel):
    """Pre-run estimate for the cost gate."""

    matched_rows: int
    # Total AI calls: matched_rows (structured) or matched_rows × #outputs.
    est_calls: int = 0
    provider_code: str | None = None
    model: str | None = None
    # Best-effort upper-bound cost; null when the model has no configured rate.
    est_cost_usd: float | None = None
    est_input_tokens_avg: int | None = None
    provider_configured: bool = False

File: frontend/lib/aiHelper.ts
import { api } from "./api";

/** Legacy v1 run-level mode (still returned for old runs). */
export type AiHelperMode = "read" | "edit";
/** v1.1 fan-out engine. */
export type AiHelperEngine = "structured" | "per_output";
/** Per-output mode: produce new content, or rewrite the column in place. */
export type AiHelperOutputMode = "write" | "edit";
export type AiHelperStatus =
  | "queued"
  | "running"
  | "cancelled"
  | "done"
  | "failed";

export interface AiHelperOutput {
  column_id: number;
  mode: AiHelperOutputMode;
  /** JSON key routed to this column (structured engine). */
  key: string;
  /** This output's own prompt (per_output engine). */
  prompt: string;
  /** Column name (populated by the backend on read). */
  name?: string;
}

export interface AiHelperCell {
  id: number;
  row_id: number;
  row_position: number;
  column_id: number;
  state: "pending" | "done" | "failed" | "skipped";
  old_value: string | null;
  new_value: string | null;
  error: string | null;
}

export interface AiHelperRunDetail {
  id: number;
  table_id: number;
  status: AiHelperStatus;
  mode: AiHelperMode;
  engine: AiHelperEngine;
  name: string | null;
  target_column_id: number | null;
  total: number;
  done: number;
  failed: number;
  skipped: number;
  reverted_at: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  prompt: string;
  variable_map: Record<string, number>;
  outputs: AiHelperOutput[];
  provider_code: string | null;
  model: string | null;
  input_scope: string;
  input_pct: number | null;
  error: string | null;
  items: AiHelperCell[];
  items_total: number;
  items_page: number;
  items_page_size: number;
}

/** Light run row for the tool-page history list (no cells). */
export interface AiHelperRunListItem {
  id: number;
  status: AiHelperStatus;
  engine: AiHelperEngine;
  mode: AiHelperMode;
  name: string | null;
  total: number;
  done: number;
  failed: number;
  skipped: number;
  reverted_at: string | null;
  created_at: string;
}

export interface AiHelperPreview {
  matched_rows: number;
  est_calls: number;
  provider_code: string | null;
  model: string | null;
  est_cost_usd: number | null;
  est_input_tokens_avg: number | null;
  provider_configured: boolean;
}

export interface AiHelperRunCreate {
  engine: AiHelperEngine;
  prompt: string;
  prompt_id?: number | null;
  name?: string | null;
  variable_map: Record<string, number>;
  outputs: AiHelperOutput[];
  provider_code?: string | null;
  model?: string | null;
  max_output_tokens?: number | null;
  input_scope: "full" | "first_pct";
  input_pct?: number | null;
  slice_column_id?: number | null;
  row_ids: number[];
}

/** Slug a column name into a JSON key: lowercase, non-alphanumerics → "_". */
export function slugifyKey(name: string): string {
  const base = (name || "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "");
  return base || "field";
}

/** Slug `name`, suffixing _2, _3… so it doesn't collide with `taken`. */
export function uniqueKey(name: string, taken: Set<string>): string {
  const base = slugifyKey(name);
  if (!taken.has(base)) return base;
  let i = 2;
  while (taken.has(`${base}_${i}`)) i += 1;
  return `${base}_${i}`;
}

/** Same {{var}} rule as the backend (prompts.py): letters/digits/_-.  + internal
 *  spaces, outer whitespace trimmed, internal runs collapsed. */
const _VAR = /\{\{\s*([A-Za-z_][\w.\- ]*?)\s*\}\}/g;

export function extractVariables(prompt: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  let m: RegExpExecArray | null;
  _VAR.lastIndex = 0;
  while ((m = _VAR.exec(prompt)) !== null) {
    const name = m[1].replace(/\s+/g, " ").trim();
    if (name && !seen.has(name)) {
      seen.add(name);
      out.push(name);
    }
  }
  return out;
}

export function createAiHelperRun(
  tableId: number,
  payload: AiHelperRunCreate,
): Promise<AiHelperRunDetail> {
  return api<AiHelperRunDetail>(`/library/tables/${tableId}/ai-helper`, {
    method: "POST",
    body: payload,
  });
}

export function previewAiHelperRun(
  tableId: number,
  payload: AiHelperRunCreate,
): Promise<AiHelperPreview> {
  return api<AiHelperPreview>(`/library/tables/${tableId}/ai-helper/preview`, {
    method: "POST",
    body: payload,
  });
}

export function listAiHelperRuns(
  tableId: number,
): Promise<AiHelperRunListItem[]> {
  return api<AiHelperRunListItem[]>(
    `/library/tables/${tableId}/ai-helper-runs`,
  );
}

export function getAiHelperRun(
  runId: number,
  page = 1,
  pageSize = 50,
): Promise<AiHelperRunDetail> {
  return api<AiHelperRunDetail>(
    `/library/ai-helper-runs/${runId}?page=${page}&page_size=${pageSize}`,
  );
}

export function cancelAiHelperRun(runId: number): Promise<AiHelperRunDetail> {
  return api<AiHelperRunDetail>(`/library/ai-helper-runs/${runId}/cancel`, {
    method: "POST",
  });
}

export function retryFailedAiHelperRun(
  runId: number,
): Promise<AiHelperRunDetail> {
  return api<AiHelperRunDetail>(`/library/ai-helper-runs/${runId}/retry-failed`, {
    method: "POST",
  });
}

export function revertAiHelperRun(runId: number): Promise<AiHelperRunDetail> {
  return api<AiHelperRunDetail>(`/library/ai-helper-runs/${runId}/revert`, {
    method: "POST",
  });
}
